staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,616 @@
+// which scales inversely to the # of active flows
+func (qs *queueSetImpl) synctime() {
+	realNow := qs.clk.Now()
+	timesincelast := realNow.Sub(qs.lastRealTime).Seconds()
+	qs.lastRealTime = realNow
+	qs.vt += timesincelast * qs.getvirtualtimeratio()
+}
+
+func (qs *queueSetImpl) lockAndSyncTime() {
+	qs.lock.Lock()
+	qs.synctime()
+}
+
+// isEmpty is a convenience method that returns 'true' when all of the queues
+// in an queueSetImpl have no packets (and is "empty")
+func (qs *queueSetImpl) isEmpty() bool {
+	return qs.numRequestsEnqueued == 0
+}
+
+// timeoutOldRequestsAndRejectOrEnqueue encapsulates the logic required
+// to validate and enqueue a request for the queueSetImpl/QueueSet:
+// 1) Start with shuffle sharding, to pick a queue.
+// 2) Reject old requests that have been waiting too long
+// 3) Reject current request if there is not enough concurrency shares and
+// we are at max queue length
+// 4) If not rejected, create a request and enqueue
+// returns the enqueud request on a successful enqueue
+// returns nil in the case that there is no available concurrency or
+// the queuelengthlimit has been reached
+// NOTE: This function must only be called with the QueueSet locked
+func (qs *queueSetImpl) timeoutOldRequestsAndRejectOrEnqueue(hashValue uint64) *fq.Request {
+
+	//	Start with the shuffle sharding, to pick a queue.
+	queueIdx := qs.chooseQueueIdx(hashValue)
+	queue := qs.queues[queueIdx]
+	// The next step is the logic to reject requests that have been waiting too long
+	qs.removeTimedOutRequestsFromQueue(queue)
+	// NOTE: currently timeout is only checked for each new request.  This means that there can be
+	// requests that are in the queue longer than the timeout if there are no new requests
+	// We think this is a fine tradeoff
+
+	// Create a request and enqueue
+	req := &fq.Request{
+		DequeueChannel: make(chan bool, 1),
+		EnqueueTime:    qs.clk.Now(),
+		Queue:          queue,
+	}
+	if ok := qs.rejectOrEnqueue(req); !ok {
+		return nil
+	}
+	metrics.ObserveQueueLength(qs.name, len(queue.Requests))
+	return req
+
+}
+
+// removeTimedOutRequestsFromQueue rejects old requests that have been enqueued
+// past the requestWaitLimit
+func (qs *queueSetImpl) removeTimedOutRequestsFromQueue(queue *fq.Queue) {
+	timeoutIdx := -1
+	now := qs.clk.Now()
+	reqs := queue.Requests
+	// reqs are sorted oldest -> newest
+	// can short circuit loop (break) if oldest packets are not timing out
+	// as newer packets also will not have timed out
+
+	// now - requestWaitLimit = waitLimit
+	waitLimit := now.Add(-qs.requestWaitLimit)
+	for i, req := range reqs {
+		if waitLimit.After(req.EnqueueTime) {
+			qs.counter.Add(1)
+			req.DequeueChannel <- false
+			close(req.DequeueChannel)
+			// get idx for timed out packets
+			timeoutIdx = i
+
+		} else {
+			break
+		}
+	}
+	// remove timed out packets from queue
+	if timeoutIdx != -1 {
+		// timeoutIdx + 1 to remove the last timeout req
+		removeIdx := timeoutIdx + 1
+
```

Comment:
nit: I feel like this blank is telling me that the block below is somehow a chunk of work that is useful to understand separately from the one above, but it doesn't appear to be the case?

Reply:
Removed the blank